### DISCUSSION TITLE
How to generate the correct type associations in a schema.objectType?

### DISCUSSION THREAD

**Discussion by TheoMer** [prisma/prisma#3167] [Q&A] [answered]
So in Prisma 1 datamodel.graphql I have defined:

```graphql
type ItemVariants {
  quantity: Int! @default(value: 1)
  user: User!
  cartitems: [CartItem!]! @relation(name: "CartItemsOfItemVariants" onDelete: CASCADE) @scalarList(strategy: RELATION)
  item: Item! @relation(name: "ItemVariantsOfItem") #relationship to Item. 
  size: Size!
  color: Color!
}
```

which generates the following schema type:

```graphql
type ItemVariants {
  quantity: Int!
  user: User!
  cartitems(where: CartItemWhereInput, orderBy: CartItemOrderByInput, skip: Int, after: String, before: String, first: Int, last: Int): [CartItem!]
  item: Item!
  size: Size!
  color: Color!
}
```

In Prisma 2 I have defined an objectType as follows:

```ts
  schema.objectType({
    name: 'ItemVariants',
    definition(t) {
      t.model.quantity() <!--- Do I specify an @default parameter here?
      t.model.user()
      t.model.cartitems({
        filtering: true,
        ordering: true,
      })
      t.model.item()
      t.model.size()
      t.model.color()
    },
  })
```

which generates the following type:

```graphql
type ItemVariants {
  cartitems(after: CartItemWhereUniqueInput, before: CartItemWhereUniqueInput, first: Int, last: Int, orderBy: CartItemOrderByInput, where: CartItemWhereInput): [CartItem!]!
  color: String <!--- This is incorrect
  quantity: Int!
  size: String <!--- This is incorrect
  user: String <!--- This is incorrect
}
```

But what I need is:

```graphql
type ItemVariants {
  ........
  color: Color!
  quantity: Int!
  size: Size!
  user: User!
}
```

How do I correctly specify this in the objectType?

**Comment by TheoMer** [comment, marked as answer]
So it appears that the main issue was that any arrays initially specified in Prisma 1 as `user: User` are being converted, after introspection to `User: User`

schema.prisma
```prisma
model ItemVariants{
  user            String?
  User            User?    @relation(fields: [user], references: [id])
}
```

My referencing `t.model.user()` in objectType of course then references `user: String?` Swapping the cases around resolved the type issue. But the type definition in nexus.graphql is still not showing the field as required

```graphql
type ItemVariants {
  user: User
}
```

Shouldn't that be picked up from the model definition?

**Reply by nikolasburk** [reply]
Hey @TheoMer 👋 can you try to use the uppercase `User` field to your `schema.objectType` definition as well:

```ts
  schema.objectType({
    name: 'ItemVariants',
    definition(t) {
      t.model.quantity() <!--- Do I specify an @default parameter here?
      t.model.user()
      t.model.User()
      t.model.cartitems({
        filtering: true,
        ordering: true,
      })
      t.model.item()
      t.model.size()
      t.model.color()
    },
  })
```

Relations in Prisma 2 work a bit differently than in Prisma 1, you can read more about this topic in the docs [here](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations).

Also, are you currently upgrading from Prisma 1? If so, are you using the [upgrade documentation](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/how-to-upgrade) for that process?

**Reply by TheoMer** [reply]
Hey @nikolasburk many thanks for the reply. Yes, I'm using the upgrade documentation you reference.

What parameter do I add to `t.model.User()` to make it required?

```
type ItemVariants {
  user: User!
}
```

I tried:

```
  schema.objectType({
    name: 'ItemVariants',
    definition(t) {
      t.model.User({ nullable: false })
    },
  })
```

but this throws an `Argument of type '{ nullable: boolean; }' is not assignable to parameter of type 'BaseRelationOptions<"ItemVariants", "User", undefined, "User">'.` error in VSCode

```
      t.list.field('User', {
        type: 'User',
        nullable: false,
      })
```

also threw an error in VSCode.

**Reply by ryands17** [reply]
Nexus automatically adds required or optional based on the model that you have defined.
The model defined above

```prisma
model ItemVariants{
  user            String?
  User            User?    @relation(fields: [user], references: [id])
}
```

Has User type as optional in ItemVriants (the `?`)

Remove that and after generating Prisma Client, Nexus will automatically add it as required:

```prisma
model ItemVariants{
  user            String
  User            User    @relation(fields: [user], references: [id])
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
